stustapay/core/service/order.py

```diff
@@ -65,90 +93,115 @@ async def create_order(
             count,
             order_id,
         )
+        order = await self._fetch_order(conn=conn, order_id=order_id)
+        if order is None:
+            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR)
 
-        return OrderID(order_id)
+        # NOW book the order, or fail
+        if new_order.order_type == ORDER_TYPE_SALE:
+            await self._book_sale_order(conn=conn, order=order, customer=customer_account)
+        else:
+            raise NotImplementedError()
 
-    @with_db_transaction
-    async def show_order(self, *, conn: asyncpg.Connection, order_id: int) -> Optional[Order]:
-        """
-        get all info about an order.
-        """
-        row = await conn.fetchrow("select * from order_value where id = $1", order_id)
-        if row is None:
-            return None
+        await self._finish_order(conn=conn, order=order)
 
-        db_line_items = await conn.fetch(
-            "select * from lineitem join product on product_id = id where order_id = $1", order_id
-        )
-        line_items = [LineItem.from_db(row) for row in db_line_items]
-        return Order.from_db(row, line_items)
+        new_balance = await conn.fetchval("select balance from account where id = $1", customer_account.id)
+        return CompletedOrder(order_id, order_uuid, old_balance=customer_account.balance, new_balance=new_balance)
 
-    @with_db_transaction
-    async def order_payment_info(self, *, conn: asyncpg.Connection, order_id: int):
+    async def _book_sale_order(self, *, conn: asyncpg.Connection, order: Order, customer: Account) -> float:
         """
-        try to pay a pending order, so one can see the available payment options.
+        The customer wants to buy same wares, like Beer.
+        It is checked if enough funds are available and books the results
+        returns the new balance of the customer account
         """
-        del order_id, conn
-        raise NotImplementedError()
+        assert order.order_type == ORDER_TYPE_SALE
+        if customer.balance < order.value_sum:
+            raise HTTPException(
+                status_code=status.HTTP_402_PAYMENT_REQUIRED, detail="Not enough funds on customer account"
```

Comment:
instead of funny http exceptions and error codes, we should rather define an error struct with custom error codes that the frontend can properly decode and then display as error message.

Reply:
and: when this fails, do we wanna completely discard the new order? because the dbtransaction fails that way.

Reply:
Yes, the error codes should be better defined. We can ask @mikonse what the best way is. We could do HTTP 400 for all fails and specify an internal error code, why the order cannot be accepted.